Update instructions for running Rsync as root for PSA-enabled clusters

RichardHoch opened the pull request:
MTC 1.7.5+,  OCP 4.6+

Resolves https://bugzilla.redhat.com/show_bug.cgi?id=2136284 by adding https://github.com/migtools/mig-operator/pull/856 to the downstream documentation.

Preview:

stillalearner commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+
+.Procedure
+
+* Prior to migration, if you are running workloads as root on the source side of a migration, set the `enforce`, `audit`, and `warn` labels on the source side of the Rsync pod as follows:
```
"set the `enforce`, `audit`, and `warn` labels on the source side of the Rsync pod as follows"

This line doesn't make it very clear about where to set those labels.

It should be clearly mentioned that the labels to be set on the **source cluster namespace** running the workloads.

RichardHoch replied:
Fixed.

stillalearner commented on modules/migration-rsync-migration-controller-root-non-root.adoc:
```diff
@@ -0,0 +1,28 @@
++
+This configuration will apply to all future migrations.
+
+. To run Rysync as non-root, which is the default setting, configure the `MigrationController` CR as above, except set `migration_rsync_privileged` to `false`.
```
"By default, Rsync run as non-root."

RichardHoch replied:
Fixed.

stillalearner commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+
+.Procedure
+
+* Prior to migration, if you are running workloads as root on the source side of a migration, set the `enforce`, `audit`, and `warn` labels on the source cluster namespaces of the workspaces:
```
"workspaces"

you mean workloads?

stillalearner reviewed (approved):
lgtm

RichardHoch commented:
/OADP
/peer-review-needed

RichardHoch commented:
/label OADP
/label peer-review-needed

aireilly commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+[id="migration-rsync-manually-overriding_{context}"]
+== Manually overriding default non-root operation for data transfer
+
+Although running Rsync pods as non-root user works in most cases, data transfer might fail when when you run workloads as root user on the source side. {mtc-short} provides two ways to manually override default non-root operation for data transfer: configuring all migrations to run an Rsync pod as root on the destination cluster for all migrations or to run an Rsync pod as root on the destination ...
```
Nit: Use a bulleted list for clarity

```suggestion
Although running Rsync pods as non-root user works in most cases, data transfer might fail when when you run workloads as root user on the source side. {mtc-short} provides two ways to manually override default non-root operation for data transfer: 
* Configure all migrations to run an Rsync pod as root on the destination cluster for all migrations
* Run an Rsync pod as root on the destination cluster per migration
```

RichardHoch replied:
@aireilly I see what happened here -- I copy-pasted the upstream destription into one file, which I at frst thought would be a single module, then I edited the text. At a certain point it became obvious that this this file should be broken into modules, which I did, working from my large moule. I should've simply gotten rid of teh latge module after (note it's not called by any assembly). I've deleted it, but made the changes you suggest in the componenet modules that I created for thsi PR.

RichardHoch replied:
Bulleted list -- great idea. Done.

aireilly commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+
+.Procedure
+
+* Prior to migration, if you are running workloads as root on the source side of a migration, set the `enforce`, `audit`, and `warn` labels on the source cluster namespaces of the workloads:
```
Nit: since this is a procedure, it would be nice to add the step required to actually apply the label, eg., something like:

```cmd
oc label pod example-pod mylabel=myvalue
```

RichardHoch replied:
Done.

aireilly commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+
+[IMPORTANT]
+====
+If you do not set these labels, Rsync will automatically fallback to non-root user even if you configure `MigrationController` or `MigMigration` as described below.
```
ISG: don't use prepositions like "below" where possible. https://www.ibm.com/docs/en/ibm-style?topic=word-usage

```suggestion
If you do not set these labels, Rsync will automatically fallback to non-root user even if you configure `MigrationController` or `MigMigration` as described in "Configuring the MigMigration CR as root or non-root per migration".
```

aireilly commented on modules/migration-rsync-root-non-root.adoc:
```diff
@@ -0,0 +1,105 @@
+
+To guarantee successful data transfer in all environments, {mtc-full} ({mtc-short}) 1.7.5 introduced changes in Rsync pods, including running Rsync pods as non-root user by default. This ensures that data transfer is possible even for workloads that do not necessarily require higher privileges. This change was made because it is best to run workloads with lowest level of privileges possible.
+
+== Manually overriding default non-root operation for data transfer
```
Is this content copy/pasted verbatim from modules/migration-rsync-manually-overriding.adoc? Consider applying mod docs principles to not duplicate the content.

aireilly commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,32 @@
+
+:_content-type: PROCEDURE
+[id="migration-rsync-manually-overriding_{context}"]
+== Manually overriding default non-root operation for data transfer
```
```suggestion
= Manually overriding default non-root operation for data transfer
```

aireilly reviewed (commented):
Looks good! Well written and authoritative. 

I have a general question regarding modules/migration-rsync-root-non-root.adoc which seems to duplicate content verbatim from another file in the PR.

aireilly commented:
/label peer-review-done
/remove-label peer-review-in-progress
/remove-label peer-review-needed

RichardHoch commented:
It would be best if @aireilly re-read this PR, but I leave that to you. GA is 15 December, so time is the most important factor. Thanks!

aireilly commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,42 @@
++
+[source,terminal]
+----
+oc label pod <PodName> <LabelName>=<LabelValue>
```
OCPs docs style for user replaced variables is to wrap any replaceable values in angle brackets (<>) with the required command parameter, using underscores (_) between words as necessary for legibility.

```suggestion
oc label pod <PodName> <label_name>=<label_value>
```

aireilly reviewed (commented):
Thanks Richard. One small nit, otherwise LGTM!

aireilly commented:
/label peer-review-done
/remove-label peer-review-in-progress
/remove-label peer-review-needed

aireilly commented:
/lgtm

RichardHoch commented:
/OADP
/merge-review-needed

RichardHoch commented:
GA is tomorrow, 15 December.

bergerhoffer commented:
@RichardHoch Are you saying that this PR should not be merged until tomorrow December 15?

michaelryanpeter commented:
@RichardHoch I have some concerns about where this PR needs to be cherrypicked to. OCP 4.6 is end of life, so I am assuming you mean 4.8+?

That said, pod security admissions were introduced in OpenShift 4.11. So while the procedure for running as root/non-root might apply to OpenShift 4.8+, my understanding is that the part about PSAs would only apply to 4.11+. 

@stlaz and @bparees Would you PTAL at this PR?
cc: @bergerhoffer

kalexand-rh commented:
I don't think we should merge this until we hear back from @stlaz and/or @bparees.

bparees commented on migration_toolkit_for_containers/installing-mtc-restricted.adoc:
```diff
@@ -37,6 +37,24 @@ include::modules/migration-configuring-proxies.adoc[leveloffset=+2]
+
+OpenShift environments have the `PodSecurityAdmission` controller enabled by default. This controller requires cluster administrators to enforce Pod Security Standards by means of namespace labels. All workloads in the cluster are expected to run one of the following Pod Security Standard levels: `Privileged`, `Baseline` or `Restricted`. Every cluster has its own default policy set.
+
+To guarantee successful data transfer in all environments, {mtc-full} ({mtc-short}) 1.7.5 introduced changes in Rsync pods, including running Rsync pods as non-root user by default. This ensures that data transfer is possible even for workloads that do not necessarily require higher privileges. This change was made because it is best to run workloads with lowest level of privileges possible.
```
```suggestion
To guarantee successful data transfer in all environments, {mtc-full} ({mtc-short}) 1.7.5 introduced changes in Rsync pods, including running Rsync pods as non-root user by default. This ensures that data transfer is possible even for workloads that do not necessarily have access to higher privileges. This change was made because it is best to run workloads with lowest level of privileges possible.
```

?

bparees replied:
also it's not clear....does this mean rsync can run using `restricted` permissions or not?  running non-root is one aspect, but it doesn't tell me whether any other privileges are still needed that prevent running rsync pods in a restricted NS.

RichardHoch replied:
@stillalearner -- are other privileges needed?

bparees commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,42 @@
+* Configure all migrations to run an Rsync pod as root on the destination cluster for all migrations.
+* Run an Rsync pod as root on the destination cluster per migration.
+
+In both cases, you must set the following labels on the source side of any namespaces that are running workloads with higher privileges prior to migration: `enforce`, `audit`, and `warn.`
```
why do you need to set audit and warn labels?  they don't actually prevent/allow anything.

bparees replied:
also "set them" is meaningless without saying what they need to be set to.  So this should say "must set the following labels to `privileged`" (but again, only the enforce label really matters)

stillalearner replied:
@pranavgaikwad ^

stlaz replied:
Just setting them also won't help you if you're running in a namespace that has these auto-labelled.

bparees commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,42 @@
+pod-security.kubernetes.io/warn: privileged
+----
++
+These labels set an exception to the destination cluster's default security policy and allow Rsync pods to run with higher privileges than the enforced defaults.
```
most likely you also need to disable PSA label syncing in the NS, otherwise the label being set here is going to get reverted by the PSA label syncer.

@stlaz can provide the details around this.

stlaz replied:
Exactly. We've got docs for it - https://docs.openshift.com/container-platform/4.11/authentication/understanding-and-managing-pod-security-admission.html#security-context-constraints-psa-opting_understanding-and-managing-pod-security-admission

RichardHoch replied:
Substitution made.

RichardHoch commented:
@kalexand-rh: I don't know enough to answer these questions...calling in the OADP dev and QE team members who reviewed this PR before i submitted it for OCP doc review.

@pranavgaikwad @stillalearner please answer the questions raised in the last 24 hours -- starts with question about which version of OCP this applies to and moves into questions about PSA.  Thanks!

stillalearner commented:
@RichardHoch This PR applies to the OCP version 4.11 and above only.

Also, the changes made in MTC were already released with 1.7.5, and this PR can be merged.

RichardHoch commented:
@michaelryanpeter : Good atch -- you're right -- the PR is for OCP 4.11+ 
@kalexand-rh and @bparees -- please see Sachin Singla's comments [OADP QE, "stillalearner"]

stlaz commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,42 @@
++
+[source,terminal]
+----
+oc label pod <pod_name> <label_name>=<label_value>
```
this is labelling a pod, should be labelling a namespace

stlaz commented on modules/migration-rsync-manually-overriding.adoc:
```diff
@@ -0,0 +1,42 @@
++
+[source,yaml]
+----
+pod-security.kubernetes.io/enforce: privileged
```
IIRC running a workload as root still puts you on the baseline level so setting the labels to "privileged" would actually not be advised.

RichardHoch commented:
@pranavgaikwad Please review the comments made in the last day.

@pranavgaikwad @stillalearner @stlaz  -- it looks the problems are with the "migration-rsync-manually-overriding.adoc" file.  Would it make sense to replace the file with a reference to https://docs.openshift.com/container-platform/4.11/authentication/understanding-and-managing-pod-security-admission.html#security-context-constraints-psa-opting_understanding-and-managing-pod-security-admissionv? If not, I need a bit more clarity about what shuld be in "migration-rsync-manually-overriding.adoc" . Thanks.

pranavgaikwad commented:
@RichardHoch I think it makes perfect sense to put reference to that doc instead of explaining the details inline.

stillalearner commented:
@RichardHoch sounds good to me as well.

RichardHoch commented:
@kalexand-rh @stlaz   I removed the module that was getting all the comments and replaced it with a reference as stlaz suggested. I hope that resolves all the issues. If so, please merge the PR, because the version was GA last week. If not, let me know...

RichardHoch commented:
/label OADP
/label merge-review-needed

gaurav-nelson commented:
/cherrypick enterprise-4.11
